zipline/assets/_securities.pyx

```diff
@@ -0,0 +1,189 @@
+#
+# Copyright 2015 Quantopian, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Cythonized Security object.
+"""
+cimport cython
+
+import numpy as np
+cimport numpy as np
+
+
+# IMPORTANT NOTE: You must change this template if you change
+# Security.__reduce__, or else we'll attempt to unpickle an old version of this
+# class
+CACHE_FILE_TEMPLATE = '/tmp/.%s-%s.v3.cache'
+
+
+cdef class Security:
+
+    cdef readonly int sid
+    # Cached hash of self.sid
+    cdef int sid_hash
+
+    cdef readonly object symbol
+    cdef readonly object security_name
+
+    # TODO: Maybe declare as pandas Timestamp?
+    cdef readonly object start_date
+    cdef readonly object end_date
+    cdef public object first_traded
+
+    cdef readonly object exchange
+
+    def __cinit__(self,
+                  int sid, # sid is required
+                  object symbol="",
+                  object security_name="",
+                  object start_date=None,
+                  object end_date=None,
+                  object first_traded=None,
+                  object exchange=""):
+
+        self.sid           = sid
+        self.sid_hash = hash(sid)
+        self.symbol        = symbol
```

Comment:
Something that I had not considered until reading through is that some of these extra fields are extraneous in a pure Zipline context. We could remove the 'extra' fields and have an internal Security class that inherits from this one, with the internal class adding the display related fields.

(That internal class could also handle the custom caching etc.)

However, that is more of an open question, and wouldn't say we would need to block merge on it.